### ISSUE TITLE
New version of Fortigate not supported

### ISSUE THREAD

**Issue by kwbr** [adrienverge/openfortivpn#774]
My company uses a new version of the Fortigate. openfortivpn stopped working with this version. Some issues I found:

```
DEBUG:  http_receive:                                                                                                                                                                                           
HTTP/1.1 200 OK                                                                                                                                                                                                 
Date: Fri, 21 Aug 2020 09:52:32 GMT                                                                     
Server: xxxxxxxx-xxxxx                                                                                                                                                                                          
Set-Cookie:  SVPNCOOKIE=; path=/; expires=Sun, 11 Mar 1984 12:00:00 GMT; secure; httponly;
Set-Cookie: SVPNNETWORKCOOKIE=; path=/remote/network; expires=Sun, 11 Mar 1984 12:00:00 GMT; secure; httponly
Set-Cookie: SVPNTMPCOOKIE=Q4tKTm4cVlSjXJwBCjs93tjKa0UtbN7h22oZecuAN3TZL2AK7vA9310XnikrAwZT45ypILXfWI6EVJi467wrK4TLeor2JLDGsvwIliWXV+iP6SZK5vhctpeAcnq99JOEs2vuSudZfbldsY4x9P05NA7ZhRAyiR28A5W8c6ekEVvPb44PDUs1NnS+Tu8GbMkvaHyEl4vU5iDGW1wCSOWsW8aJILCh3JCgXOLRtI1x/Fg77YPypFrQEDhV8uLDYj73WxASRqy8FygK; path=/remote/hostcheck_install; secure; httponly                                                                        Transfer-Encoding: chunked                                                                                                                                                                                      
Content-Type: text/plain                                                                                
X-Frame-Options: SAMEORIGIN                                                                             
Content-Security-Policy: frame-ancestors 'self'; object-src 'self'; script-src 'self' https  (null) 'unsafe-eval' 'unsafe-inline' blob:;
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Strict-Transport-Security: max-age=31536000

b3    
ret=1,redir=/remote/hostcheck_install?auth_type=16&user=<XXX>&portal=<XXX>&rip=<IP-address>&realm=
0
```

Seems the protocol has changed. There is an empty SVPNCOOKIE (there is even a space in front, so that openfortivpn can no longer find this cookie name).

My C skills are very basic. I managed to hack the SVPNTMPCOOKIE into the next request to /remote/index but get an "HTTP 403 Forbidden".

I would like to support to include this new protocol. How can I get a dump for example from the official Windows clients requests?

**Comment by DimitriPapadopoulos** [comment]
> On this page a client for Linux can be downloaded: https://www.forticlient.com/downloads (For Debian/Ubuntu based systems https://filestore.fortinet.com/forticlient/downloads/FortiClientFullVPNInstaller_6.4.0.0851.deb).

That's interesting. This page has been changed recently and the Linux VPN SSL official client has been made broadly available recently (probably July 2020). That said this is not the full version::
> _Upgrade to the full version to access additional features and receive technical support._

As suggested in https://github.com/adrienverge/openfortivpn/pull/713#issuecomment-633614514 there remain two versions:
1. The widely available FortiClient VPN 6.4.x lacks features, I suspect "host check" (a.k.a. "Compliance Enforcement"?) is missing.
2. The full-fledged FortiClient 6.4.x is only available at additional cost (with EMS as far as I can understand).

![FortClientVPN](https://user-images.githubusercontent.com/3234522/91005564-7e6ca480-e5d7-11ea-9927-3e75a8e978aa.png)

**Comment by DimitriPapadopoulos** [comment]
> I can connect from Windows and my colleagues using MacOS can also connect.

In version 6.4, Windows and macOS both support IPSec VPN, the Linux version supports only SSL VPN. So it could either issue:
* The FortiGate supports IPSec VPN only. I can't recall how a FortiGate behaves when VPN SSL is not supported so I cannot tell whether that's the case here. Have you tested whether SSL VPN does work on Windows and macOS?
* The Linux version of _FortiClientVPN_ does not seem to be supporting "host checks", even the latest v6.4.0.0851. but you should be able to mimick them using the development version of _openfortivpn_ and following instructions in #713.

**Comment by mrbaseman** [comment]
> We have tested FortiOS 6.2 and it works for us but I'm not certain we have tested 6.4 yet. Do you have a clue about the version of FortiOS?

I think we haven't tested against FortiOS 6.4.The FortiOS Version is contained in the xml output which openfortivpn shows in `-v -v` verbose mode.

**Comment by DimitriPapadopoulos** [comment]
> I think we haven't tested against FortiOS 6.4.The FortiOS Version is contained in the xml output which openfortivpn shows in `-v -v` verbose mode.

Unfortunately it looks like _openfortivpn_ doesn't get that far.

**Comment by kwbr** [comment]
I have a Charles Proxy trace file at hand showing the communication between the Windows client and the VPN EMS endpoint. But I do not want to share it with the whole Github community...

**Comment by DimitriPapadopoulos** [comment]
@kwbr Just to make sure: The communication between the Windows client and the FortiGate is using SSL VPN, not IPSec VPN, right?

**Comment by kwbr** [comment]
It's using SSL-VPN.

**Comment by DimitriPapadopoulos** [comment]
I have no experience with Charles Proxy so I cannot promise anything, but you can send confidential data to the core team (3 persons) using the mail address provided in [SECURITY.md](https://github.com/adrienverge/openfortivpn/blob/master/.github/SECURITY.md).

**Comment by m29h** [comment]
I also have a similar issue i am struggeling with since our gateway got upgraded to fortios `<fos platform='FG200E' major='6' minor='00' patch='12' build='0419' branch='0419' />` according to the debug logfile taken from an official and legit FortiClient running on Windows.

Meanwhile `openforticlient -vv` shows that the `POST /remote/logincheck` returns an `405 Method Not Allowed` HTTP error.

I went so far to set up a mitmproxy trying to log the real traffic from the working official client but funnily in this scenario the `POST /remote/logincheck` from the working 'legit' client now also returns a `405 Method Not Allowed`. I conjecture that newer fortios versions (or at least the particular config I am dealing with) check the client certificates. The official client uses the files `fortisslclient.crt` and `fortisslclient.key` as a SSL client authentication (which also shows in the official logs). These are **not** user/customer specific files but files they  seem to be identical in every forticlient installer. Even though I have access to these files i cannot trivially use them to authenticate my mitmproxy to the real server as I am missing the password to unlock the `fortisslclient.key` private key file.

Obviously this all is just some strange security by obscurity as the key unlock password to `fortisslclient.key` must be somewhere stored or hardcoded in FortiClient but after all I am not a pentester but just a normal user that wants a nice and working linux client. Is Fortinet really making our lives so hard? Is anyone here encountering a similar problem or has any other thoughts on my observations?

**Comment by DimitriPapadopoulos** [comment]
I also have access to a 6.* gateway, but I don't see this behaviour. As you write, it might be specific to some configurations.

**Comment by DimitriPapadopoulos** [comment]
I cannot find `fortisslclient.crt` and `fortisslclient.key` files in FortiClient 7.0.0 - but I do see them  in the files of an older FortiClient 4.4.

**Comment by DimitriPapadopoulos** [comment]
@kwbr Can you confirm it works out-of-the-box   with the latest FortiClient VPN 7.0.0 available from the new download page?
https://www.fortinet.com/support/product-downloads#vpn

For Ubuntu, the current package is:
https://filestore.fortinet.com/forticlient/downloads/forticlient_vpn_7.0.0.0018_amd64.deb

**Comment by DimitriPapadopoulos** [comment]
@m29h Does it work with the official FortiClient from the same Linux machine?

**Comment by m29h** [comment]
@DimitriPapadopoulos No, I cannot get the 7.0.0 from the site to work. I have a user certificate pem file that I used to pass to openfoirtivpn via the --user-cert and --user-key arguments. This functionality seems to be missing in the publicly available 7.0.0 version. (and also e.g. the `forticlient-vpn`package in [AUR](https://aur.archlinux.org/packages/forticlient-vpn/) That is one of the reasons why i liked to use openfortivpn in the first place (until it stopped working).

I also fount the same (old) fortisslclient.crt/key that you have fround but that is expired in May 2020. I can confirm that this old key is not protected by a password. However the new file is.

The fortisslclient.crt that comes bundled with the working official windows client is very similar but newer and expires only in 2038. It looks like the following (it looks to be very generic but i nevertheless redacted some exact dates of the cert here in the public forum.)
```
        Issuer: C=US, ST=California, L=Sunnyvale, O=Fortinet, OU=Certificate Authority, CN=support/emailAddress=support@fortinet.com
        Validity
            Not Before: xxx xx xx:xx:xx 2017 GMT
            Not After : xxx xx xx:xx:xx 2038 GMT
        Subject: C=US, ST=California, L=Sunnyvale, O=Fortinet, OU=FortiClient, CN=FortiClient-SSLVPN/emailAddress=support@fortinet.com

```

**Comment by DimitriPapadopoulos** [comment]
@m29h In your case, neither _openfortivpn_, not _FortiGate_ work on Linux. The Windows _FortiClient_ works and I'm wondering whether support for the Hostcheck feature might explain it. See #713. Hostcheck support in _FortiClient_ might depend on its exact version, for example on Linux you're probably using ([FortiClient VPN only](https://www.fortinet.com/support/product-downloads#vpn) and on Windows you might be using a different version such as (ZTNA Edition)[https://www.fortinet.com/support/product-downloads#ztna]. I'm not sure how to check the exact version.

Suggestions:
- Retrieve the UA (user agent) string and other information required by the Hostcheck feature from the Windows client as explained in https://github.com/adrienverge/openfortivpn/pull/713#issuecomment-627979636.
- Feed that information to the Fortigate using these options: `user-agent`, `hostcheck`, `check-virtual-desktop`.

Actually you might even try with the values provided in https://github.com/adrienverge/openfortivpn/pull/713#issue-417312124.

**Comment by kwbr** [comment]
> @kwbr Does it work with the latest FortiClient VPN 7.0.0 available from the new download page?
> https://www.fortinet.com/support/product-downloads#vpn

Yes this works correctly now. I did not try a new version since I opened the ticket. Will try openfortovpn later.

**Comment by m29h** [comment]
@DimitriPapadopoulos Unfortunately that does not help me. I played with hostcheck before but that did not lead to anything. 
When i look at the traffic from the working windows client through a man-in-the-middle proxy I get:
`Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1) [SV{v=01.01;}]`
Note that the windows client first polls the /remote/info endpoint which returns an XML. This XML contains 
```
<?xml version='1.0' encoding='utf-8'?><info><api encmethod='0' salt='xxxxxxxx' remoteauthtimeout='10' f='1f' /></info>
```
Next the windows client polls the /remote/login which gives a normal HTML answer. Not surch what this is used for.

The salt is then used to obfuscate the call to /remote/logincheck which instead of the HTTP requests header just submits a string  in the form like "enc=0123456789ABCDEF....." with total 200 HEX digits.
If i block the call to /remote/info in my mitmproxy then the official windows client falls back to a plain text in similar form used also by openfortivpn. More specifically the form is: "username=xxxx&credential=xxxx&just_logged_in=1&redir=%2Fremote%2Findex&ajax=1" even though openfortivpn does not set the just_logged_in. I dont think this is of any relevance in this case.

There is no sign my windows client performs any hostcheck. It just fails at this point with HTTP:405
I stay with my interpretation that while i can override my client to accept the man-in-the-middle proxy with "do not warn invalid server certificate", the server has also some layer of client authentication that already kicks in in this early stage (via fortisslclient.crt/key) and thus refuses the logincheck with HTTP:405 at that point.

**Comment by dogasantos** [comment]
+1
still happening with current code

**Comment by DimitriPapadopoulos** [comment]
@dogasantos **What** is still happening? This issue has become too complex, there are too many reports and it is unclear whether they refer to the same issue.

**Comment by dogasantos** [comment]
Hi @DimitriPapadopoulos sorry.
Let me try to break the process down in a organized way:

- The debug shows that I'm able to pass the authentication phase (credentials+otp) and receive the SVPNCOOKIE. 

- Then the gateway sends a "document.location='/sslvpn/portal.html';" response but the client sends a "GET /remote/index" that produces a 403.  

-  A second attempt is made with a "GET /remote/fortisslvpn", but same 403 happens. Strangely, I get a "INFO:   Remote gateway has allocated a VPN." 

- Then a "GET /remote/fortisslvpn_xml" also produces a 403 until it finally give up with a "ERROR:  Could not get VPN configuration (HTTP status code)."

There is no hostcheck in place.

I've tried same thing but with --cookie (in case something wrong with the otp) but same thing happens. 

It seems something similar with the error messages in this thread

**Comment by realtica** [comment]
> Hi @DimitriPapadopoulos sorry. Let me try to break the process down in a organized way:
> 
> * The debug shows that I'm able to pass the authentication phase (credentials+otp) and receive the SVPNCOOKIE.
> * Then the gateway sends a "document.location='/sslvpn/portal.html';" response but the client sends a "GET /remote/index" that produces a 403.
> * A second attempt is made with a "GET /remote/fortisslvpn", but same 403 happens. Strangely, I get a "INFO:   Remote gateway has allocated a VPN."
> * Then a "GET /remote/fortisslvpn_xml" also produces a 403 until it finally give up with a "ERROR:  Could not get VPN configuration (HTTP status code)."
> 
> There is no hostcheck in place.
> 
> I've tried same thing but with --cookie (in case something wrong with the otp) but same thing happens.
> 
> It seems something similar with the error messages in this thread

The same thing is happening to me, is it possible fix the issue?
